Create `AudioStreamWithEffects`

RoyBerardo opened the pull request:
This adds a new `AudioStream` type, ~~`AudioStreamEffect`~~ `AudioStreamWithEffects`, which applies audio bus effects directly to audio streams.

![image](https://github.com/user-attachments/assets/05f46a57-cfe3-4612-8a06-ec5b4c9c896b)

This partially addresses https://github.com/godotengine/godot-proposals/discussions/8190 ("Partially" because i believe it would be useful to apply effects to audio stream players, but that is a whole other topic)

Effects can be individually bypassed. By enabling `process_bypassed_effects`, bypassed effects will still process so bypassing and un-bypassing effects like reverb won't result in a burst of sound.

When the child stream is finished playing, this stream fades out for the duration of `tail_time`, allowing effects like reverb and delay to fade out.

This is currently a draft because I feel it would be useful to get community input on this new stream type. Some things I'd like to get some thoughts/discussion on in particular:

- Should `process_bypassed_effects` be per-effect? That would allow for more granular control but I worry about cluttering the effect list. UPDATE: Decided to make it per-effect.
- Instead of `tail_fade_curve`, should there instead just be a bool that simply enables/disables a fade on the tail? UPDATE: Decided to keep the curve.
- Would appreciate it if someone looked over the docs I made, I'm not the best at writing those.
- AudioEffect docs would need to be updated since they talk about audio effects only in the context of busses.

RoyBerardo committed:
create AudioStreamWithEffects

RoyBerardo commented:
> Should process_bypassed_effects be per-effect? That would allow for more granular control but I worry about cluttering the effect list.

Decided to make it per-effect.

> Instead of tail_fade_curve, should there instead just be a bool that simply enables/disables a fade on the tail?

Decided to keep it as a curve, no need to make it less customizable (unless it has a significant perf impact, which I don't believe it does)

Also, updated all the audio effect docs to not have them be bus-exclusive, though I'd love it if someone took a look at those.

RoyBerardo commented:
Since there doesn’t seem to be any discussion and I’ve addressed the issues I’ve put forward, I’m opening this PR :)

fire commented:
There's a proposal to move the microphone out of the Godot audio system. https://github.com/godotengine/godot/pull/108773 so I was exploring the options.

Mickeon commented:
> This also has me wondering if `AudioStreamEffect` is really the best name for this. Maybe `AudioStreamEffectChain` or `AudioStreamEffectList` would be more descriptive. Ultimately this new class isn't an effect, it's a chain/list of `AudioEffect` resources.

Indeed, just by the title alone I was rather confused at first. Anything along the lines of `AudioStreamEffectChain` should be fine, but I believe any word that would convey effects _directly applied_ to the `AudioStream` would be even better.

KoBeWi commented on doc/classes/AudioStreamPlaybackWithEffects.xml:
```diff
@@ -0,0 +1,10 @@
+		Playback instance of [AudioStreamWithEffects].
+	</brief_description>
+	<description>
+	</description>
```
The description should be filled too, but not sure what to put here.
Does this class even need to be exposed? I know other playbacks are exposed too, but since you can't do anything with it, it's rather counter-productive.

RoyBerardo replied:
Very true, not much sense having this exposed if there's nothing that can be done with it.

That being said, I have just (locally) added a way to get an `AudioEffectInstance` from an `AudioStreamPlaybackWithEffects`, so now this document actually has a reason to exist

KoBeWi commented on doc/classes/AudioStreamWithEffects.xml:
```diff
@@ -0,0 +1,101 @@
+			<return type="bool" />
+			<param index="0" name="index" type="int" />
+			<description>
+				Returns if the effect at the specified index has process_when_bypassed enabled.
```
```suggestion
				Returns [code]true[/code] if the effect at the specified index has enabled process when bypassed.
```

RoyBerardo replied:
Yeah it's definitely a mouthful of a method now that you mention it. Fortunately for us, this method (along with `process_when_bypassed`) will be removed (see https://github.com/godotengine/godot/pull/107523#issuecomment-3396772333) so we won't have to break our brains coming up with a better name.

KoBeWi commented on editor/audio/audio_stream_with_effects_editor_plugin.cpp:
Pretty sure this whole plugin is unnecessary. There is `ADD_ARRAY_COUNT` that handles it.

KoBeWi commented on scene/resources/audio_stream_with_effects.h:
```diff
@@ -0,0 +1,138 @@
+
+	Ref<AudioStreamWithEffects> stream; // The effect stream this playback is instantiated from
+	Ref<AudioStreamPlayback> playback; // The playback of the child stream
+	Vector<Ref<AudioEffectInstance>> effect_instances;
```
```suggestion
	LocalVector<Ref<AudioEffectInstance>> effect_instances;
```

Mickeon reviewed (commented):
I do have more thoughts but I'll wait a bit on implementation details.

RoyBerardo commented on doc/classes/AudioStreamWithEffects.xml:
```diff
@@ -0,0 +1,101 @@
+	</brief_description>
+	<description>
+		[AudioStreamWithEffects] applies [AudioEffect]s to a child [AudioStream]. Effects are applied in the order in which they are listed.
+		[b]Note:[/b] Different effects can require vastly different amounts of processing power. Be mindful of how often intensive effects such as reverb and delay are used with your set-up.
```
> Also, as true as this is, I believe not anywhere in the class reference or manual it is stated how performance-intensive each effect is. So it's all very up to interpretation, besides the listed examples.

That's true. I wonder if it'd be useful to benchmark and document the performance intensity of each effect, and include the results (even as high/moderate/low) in the audio effects tutorial doc. If that were done, it would be nice if that and this PR were merged side-by-side.

RoyBerardo commented on doc/classes/AudioStreamWithEffects.xml:
```diff
@@ -0,0 +1,101 @@
+			<param index="1" name="enabled" type="bool" />
+			<description>
+				Sets bypass on the effect at the specified index.
+				Bypassing an effect disables the effect's influence on the sound, essentially turning it off.
```
> What's the point of toggling bypass when, if desired, one can remove and re-add the audio effect to a AudioStreamWithEffects to achieve a similar effect?

Some reasons off the top of my head:
- Better for memory management - you're not creating and destroying and creating and destroying objects.
- Allows turning off effects at an index without regard to what the effect is.
- (similar to the above point) if someone adds an effect via the effect dropdown->new->[effect], thus creating a localized effect resource just for that `AudioStreamWithEffects`, they'd need to figure out a way to hold on to it during runtime before they remove it if they want to bring it back.
- Allows setting/getting an effect at that slot while it is currently bypassed.
- Allows tweaking an effect's parameters while it is currently bypassed.
- Consistency with the bus which already includes bypassing.
- Consistency with DAWs, Wwise, Fmod, Unity, Unreal.

RoyBerardo commented on doc/classes/AudioStreamWithEffects.xml:
```diff
@@ -0,0 +1,101 @@
+		Applies effects to an audio stream.
+	</brief_description>
+	<description>
+		[AudioStreamWithEffects] applies [AudioEffect]s to a child [AudioStream]. Effects are applied in the order in which they are listed.
```
The plural `[AudioEffect]`/`[AudioEffect]s` is a bit of a head scratcher for me here. Is there a correct way to go about this that's consistent with the other docs? I like Mickeon's suggestion with that whole first sentence, but `[AudioEffect]` being singular is throwing me off.

Also I'm personally more a fan of "Effects are applied in the order in which they are listed.". Makes more sense to me. Could be modified to this though:
"The effects are applied in the order in which they are listed."

RoyBerardo commented on scene/resources/audio_stream_with_effects.cpp:
```diff
@@ -0,0 +1,434 @@
+	effect_playback->effect_instances.resize(effects.size());
+	for (int i = 0; i < effects.size(); i++) {
+		if (effects[i].effect.is_valid()) {
+			effect_playback->effect_instances.write[i] = effects[i].effect->instantiate();
```
Just updated everything to LocalVector. I originally did what you had done here, but was getting some bugs with syncing the effects. In order to allow adding/removing/reordering/replacing effects without stopping the playback, `stream->effects` and `playback->effect_instances` need to have `AudioEffect`s and `AudioEffectInstance`s paired up at the same indices. So if `stream->effects` has an `AudioEffectReverb` at index 5, `playback->effect_instances` needs to have an `AudioEffectInstanceReverb` at index 5. What this also means is that if the user has some indices with nothing set on them, `playback->effect_instances` also needs matching empty indices. Because this appends to `playback->effect_instances` instead of writing to specified indices, a mismatch happens if the stream has null entries in its effects vector when the playback is instantiated.

RoyBerardo commented:
Updated with the above suggestions, plus a couple more things.

Changes:
- Removed the editor plugin (Switched from ADD_ARRAY to ADD_ARRAY_COUNT)
- Made `effect_playback->effect_instances` a LocalVector
- Added `AudioServer` locking/unlocking to avoid thread issues
- All functions that use indexes can now use negative indexes
- Removed "Process when bypassed"
- Added `AudioStreamPlaybackWithEffects.get_effect_instance()` and `AudioStreamPlaybackWithEffects. get_effect_instance_count()` (Originally added the second function for testing/debugging, but figured it wouldn't hurt to keep it around)
- Added documentation for `AudioStreamPlaybackWithEffects`
- Looooots of documentation changes

Appreciate all the review and feedback!

KoBeWi reviewed (approved):
Looks fine functionality-wise. Needs audio team review.

AThousandShips commented on scene/resources/audio_stream_with_effects.h:
```diff
@@ -0,0 +1,139 @@
+
+	AudioServer::PlaybackType playback_type;
+
+private:
```
```suggestion
```
Redundant

RoyBerardo commented on scene/resources/audio_stream_with_effects.h:
```diff
@@ -0,0 +1,138 @@
+class AudioStreamWithEffects : public AudioStream {
+	GDCLASS(AudioStreamWithEffects, AudioStream)
+
+	AudioServer::PlaybackType playback_type;
```
I feel silly.. this variable is unused. I imagine it may have ended up in there via sneaking under the radar past early prototyping. I've removed it in the latest version.

Akosmo commented:
Adding here something we discussed in the last audio meeting:
Having this added to the engine allow us to remove the high shelf filter embedded in `_mix_step_for_channel()` in the AudioServer, which isn't the perfect approach for `AudioStreamPlayer3D`'s attenuation.

AThousandShips commented on scene/resources/audio_stream_with_effects.cpp:
```diff
@@ -0,0 +1,471 @@
+
+	base_property_helper.set_prefix("effect_");
+	base_property_helper.set_array_length_getter(&AudioStreamWithEffects::get_effect_count);
+	base_property_helper.register_property(PropertyInfo(Variant::OBJECT, "effect", PROPERTY_HINT_RESOURCE_TYPE, "AudioEffect"), default_effect.effect, &AudioStreamWithEffects::set_effect, &AudioStreamWithEffects::get_effect);
```
```suggestion
	base_property_helper.register_property(PropertyInfo(Variant::OBJECT, "effect", PROPERTY_HINT_RESOURCE_TYPE, AudioEffect::get_class_static()), default_effect.effect, &AudioStreamWithEffects::set_effect, &AudioStreamWithEffects::get_effect);
```

AThousandShips commented on scene/resources/audio_stream_with_effects.cpp:
```diff
@@ -0,0 +1,471 @@
+	ClassDB::bind_method(D_METHOD("set_tail_fade_curve", "time"), &AudioStreamWithEffects::set_tail_fade_curve);
+	ClassDB::bind_method(D_METHOD("get_tail_fade_curve"), &AudioStreamWithEffects::get_tail_fade_curve);
+
+	ADD_PROPERTY(PropertyInfo(Variant::OBJECT, "stream", PROPERTY_HINT_RESOURCE_TYPE, "AudioStream", PROPERTY_USAGE_DEFAULT), "set_stream", "get_stream");
```
```suggestion
	ADD_PROPERTY(PropertyInfo(Variant::OBJECT, "stream", PROPERTY_HINT_RESOURCE_TYPE, AudioStream::get_class_static(), PROPERTY_USAGE_DEFAULT), "set_stream", "get_stream");
```

RoyBerardo commented:
Exciting to see that we can add descriptions to array properties now! Just added those two descriptions, would appreciate some eyes on those from someone who is better at docs than I am.

goatchurchprime commented:
I've hit a snag while trying to replicate the capabilities of the AudioStreamPlayer3D that are internal to the AudioServer using this feature.  

In order to programatically make sine waves as input I need access to the AudioStreamGeneratorPlayback for the stream that is playing so I can push in the frames.

<img width="642" height="341" alt="image" src="https://github.com/user-attachments/assets/2cce39d9-7883-401d-a422-023752453f94" />

Is there any way to get to this, or is this a flaw in the use of the `AudioStreamGeneratorPlayback` object that it is only ever accessible when it is the top level stream in the AudioStreamPlayer?
 https://docs.godotengine.org/en/stable/classes/class_audiostreamgeneratorplayback.html

goatchurchprime commented:
After some use of this feature [in a project](https://github.com/goatchurchprime/AudioStreamEffects3D/tree/useaudiostreamwitheffects) to replicate and expose the hidden AudioEffects which implement the `AudioStreamPlayer3D`, I'm not as positive about this feature as I had been earlier. 

The problem is that when you wrap the `AudioStream` of an `AudioStreamPlayer` into this new `AudioStreamWithEffects`:
 * you lose access to the `AudioStreamPlayback` of the inner `AudioStream` (so you can't use `AudioStreamGenerator`), 
 *  you lose the polyphony feature where you can call `AudioStreamPlayer.play()` multiple times and the `AudioServer` spawns new simultaneously playing `AudioStreamPlaybacks` up to the quantity of `max_polyphony`.

On the other hand, this feature doesn't appear to give us any new features that can't already be implemented by inserting a new unique bus whenever we want to assign effects onto a single AudioStreamPlayer.(*)

(*) Not quite accurate, since it allows for the application of effects before the spatial attenuation applied by AudioStreamPlayer3D.  However, once the AudioStreamPlayer3D's audioeffects are exposed on a bus, you would have the ability to insert further effects in front or behind this spatialization effect.  This is mostly not a big deal since the majority of effects commute (makes no difference the order they apply), but some do not.
